PWGHF/HFC/TableProducer/correlatorDsHadrons.cxx

```diff
@@ -213,131 +470,260 @@ struct HfCorrelatorDsHadrons {
         }
         double efficiencyWeight = 1.;
         if (applyEfficiency) {
-          efficiencyWeight = 1. / efficiencyD->at(o2::analysis::findBin(binsPt, candidate1.pt()));
-        }
-
-        if (std::abs(candidate1.flagMcMatchRec()) == 1 << DecayType::DsToKKPi) {
-          // fill per-candidate distributions from Ds true candidates
-          registry.fill(HIST("hPtCandMCRec"), candidate1.pt());
-          registry.fill(HIST("hPtProng0MCRec"), candidate1.ptProng0());
-          registry.fill(HIST("hPtProng1MCRec"), candidate1.ptProng1());
-          registry.fill(HIST("hPtProng2MCRec"), candidate1.ptProng2());
-          registry.fill(HIST("hEtaMCRec"), candidate1.eta());
-          registry.fill(HIST("hPhiMCRec"), RecoDecay::constrainAngle(candidate1.phi(), -o2::constants::math::PIHalf));
-          registry.fill(HIST("hYMCRec"), yDs(candidate1));
-          registry.fill(HIST("hSelectionStatusMCRec"), candidate1.isSelDsToKKPi());
-        }
-        // fill invariant mass plots from Ds signal and background candidates
-        registry.fill(HIST("hMassDsMCRec"), invMassDsToKKPi(candidate1), efficiencyWeight);
-        if (candidate1.flagMcMatchRec() == 1 << DecayType::DsToKKPi) { // also matched as Ds
-          registry.fill(HIST("hMassDsMCRecSig"), invMassDsToKKPi(candidate1), candidate1.pt(), efficiencyWeight);
+          efficiencyWeight = 1. / efficiencyD->at(o2::analysis::findBin(binsPt, candidate.pt()));
+        }
+        auto prong0McPart = candidate.prong0_as<myTracksMc>().mcParticle_as<candDsMcGen>();
+        if (std::abs(candidate.flagMcMatchRec()) == 1 << DecayType::DsToKKPi) {
+          fillHistoMcRecSig(candidate, multiplicityV0M);
+          // KKPi
+          if (std::abs(prong0McPart.pdgCode()) == kKPlus) {
+            registry.fill(HIST("hMassDsMCRec"), invMassDsToKKPi(candidate), efficiencyWeight);
+            registry.fill(HIST("hMassDsMCRecSig"), invMassDsToKKPi(candidate), candidate.pt(), efficiencyWeight);
+            registry.fill(HIST("hMassDsVsPtMCRec"), invMassDsToKKPi(candidate), candidate.pt(), efficiencyWeight);
+            registry.fill(HIST("hSelectionStatusMCRec"), candidate.isSelDsToKKPi());
+          }
+          // PiKK
+          if (std::abs(prong0McPart.pdgCode()) == kPiPlus) {
+            registry.fill(HIST("hMassDsMCRec"), invMassDsToPiKK(candidate), efficiencyWeight);
+            registry.fill(HIST("hMassDsMCRecSig"), invMassDsToPiKK(candidate), candidate.pt(), efficiencyWeight);
+            registry.fill(HIST("hMassDsVsPtMCRec"), invMassDsToPiKK(candidate), candidate.pt(), efficiencyWeight);
+            registry.fill(HIST("hSelectionStatusMCRec"), candidate.isSelDsToPiKK());
+          }
+          // prompt
+          if (candidate.originMcRec() == RecoDecay::OriginType::Prompt) {
+            flagDsPrompt = true;
+          }
+          // non-prompt
+          if (candidate.originMcRec() == RecoDecay::OriginType::NonPrompt) {
+            flagDsPrompt = false;
+          }
```

Comment:
Hi @vkucera,
I have declared the variable flagDsPrompt outside the loop over the candidates so the variable is not created in every iteration, and then in the loop I use this structure `if` to divide the two cases prompt and non-prompt. Now I have corrected the code using an `if-else` structure instead of two `if`. I have done in this way because there are two flags and for the readability of the code so it is clear the separation. I don't understand if I should remove one condition (e.g. the Non-prompt condition) and set the value of `flagDsPrompt` to `false`  if it doesn't pass the condition of Prompt selection or to remove both conditions and do the same of `flagDsSignal`.
Thanks in advance,
Samuele

Reply:
Because you declare `flagDsPrompt` outside the loop, you have to set its value explicitly for each candidate. Since there are only two possible values of `originMcRec` (`Prompt/NonPrompt`), you can set the value on a single line in the same way you do for `flagDsSignal`.

Reply:
Ok, thanks. Now I have understood.